rollout/trafficrouting/traefik/traefik.go

```diff
@@ -0,0 +1,164 @@
+package traefik
+
+import (
+	"context"
+	"strings"
+
+	"github.com/pkg/errors"
+
+	"github.com/argoproj/argo-rollouts/pkg/apis/rollouts/v1alpha1"
+	"github.com/argoproj/argo-rollouts/utils/defaults"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/dynamic"
+)
+
+// Type holds this controller type
+const Type = "Traefik"
+
+var (
+	apiGroupToResource = map[string]string{
+		"traefik.containo.us": "traefikservices",
+	}
+)
+
+type ReconcilerConfig struct {
+	Rollout *v1alpha1.Rollout
+	Client  ClientInterface
+}
+
+type Reconciler struct {
+	Rollout *v1alpha1.Rollout
+	Client  ClientInterface
+}
+
+type ClientInterface interface {
+	Get(ctx context.Context, name string, options metav1.GetOptions, subresources ...string) (*unstructured.Unstructured, error)
+	Update(ctx context.Context, obj *unstructured.Unstructured, options metav1.UpdateOptions, subresources ...string) (*unstructured.Unstructured, error)
+}
+
+func NewReconciler(cfg ReconcilerConfig) *Reconciler {
+	reconciler := &Reconciler{
+		Rollout: cfg.Rollout,
+		Client:  cfg.Client,
+	}
+	return reconciler
+}
+
+func NewDynamicClient(di dynamic.Interface, namespace string) dynamic.ResourceInterface {
+	return di.Resource(GetMappingGVR()).Namespace(namespace)
+}
+
+func GetMappingGVR() schema.GroupVersionResource {
+	group := defaults.DefaultTraefikAPIGroup
+	parts := strings.Split(defaults.DefaultTraefikVersion, "/")
+	version := parts[len(parts)-1]
+	resourceName := apiGroupToResource[group]
+	return schema.GroupVersionResource{
+		Group:    group,
+		Version:  version,
+		Resource: resourceName,
+	}
+}
+
+func (r *Reconciler) UpdateHash(canaryHash, stableHash string, additionalDestinations ...v1alpha1.WeightDestination) error {
+	return nil
+}
+
+func (r *Reconciler) SetWeight(desiredWeight int32, additionalDestinations ...v1alpha1.WeightDestination) error {
+	ctx := context.TODO()
+	rollout := r.Rollout
+	traefikServiceName := rollout.Spec.Strategy.Canary.TrafficRouting.Traefik.TraefikServiceName
+	traefikService, err := r.Client.Get(ctx, traefikServiceName, metav1.GetOptions{})
+	if err != nil || traefikService == nil {
+		return err
```

Comment:
If `traefikService == nil` it will return a nil error for this exception. It's better to handle this error separately and check for the proper error type in case it isn't found. 
Example: `k8s.io/apimachinery/pkg/api/errors.IsNotFound(err)`

Reply:
I checked that if it cant find the needing resource it return corresponding error so I think I can remove the check traefikService